lib/rules/declaration-property-max-values/index.js

```diff
@@ -0,0 +1,71 @@
+'use strict';
+
+const matchesStringOrRegExp = require('../../utils/matchesStringOrRegExp');
+const report = require('../../utils/report');
+const ruleMessages = require('../../utils/ruleMessages');
+const valueParser = require('postcss-value-parser');
+const vendor = require('../../utils/vendor');
+const validateOptions = require('../../utils/validateOptions');
+const { isNumber } = require('../../utils/validateTypes');
+const validateObjectWithProps = require('../../utils/validateObjectWithProps');
+
+const ruleName = 'declaration-property-max-values';
+
+const messages = ruleMessages(ruleName, {
+	rejected: (property, max) =>
+		`Expected "${property}" to have no more than ${max} ${max === 1 ? 'value' : 'values'}`,
+});
+
+const meta = {
+	url: 'https://stylelint.io/user-guide/rules/list/declaration-property-max-values',
+};
+
+const isValueNode = (/** @type {valueParser.Node} */ node) => {
+	return node.type === 'word' || node.type === 'function' || node.type === 'string';
+};
+
+/** @type {import('stylelint').Rule<Record<string, number>>} */
+const rule = (primary) => {
+	return (root, result) => {
+		root.walkDecls((decl) => {
+			const validOptions = validateOptions(result, ruleName, {
+				actual: primary,
+				possible: [validateObjectWithProps(isNumber)],
+			});
+
+			if (!validOptions) {
+				return;
+			}
+
```

Comment:
[suggest] It's better to move `validateOptions` outside `walkDecls` for performance:

```suggestion
		const validOptions = validateOptions(result, ruleName, {
			actual: primary,
			possible: [validateObjectWithProps(isNumber)],
		});

		if (!validOptions) {
			return;
		}

		root.walkDecls((decl) => {
```